src/modelbench/benchmark_runner.py

```diff
@@ -0,0 +1,397 @@
+
+            response = mg_sut.translate_response(raw_request, raw_response)
+            item.sut_response = response
+        except Exception as e:
+            item.exception = e
+        return item
+
+
+class TestRunAnnotationWorker(IntermediateCachingPipe):
+
+    def __init__(self, test_run: TestRun, thread_count=1, cache_path=None):
+        super().__init__(thread_count, cache_path=cache_path)
+        self.test_run = test_run
+
+    def handle_item(self, item: TestRunItem) -> TestRunItem:
+        try:
+            if item.completion():
+                self.collect_annotations(item)
+        except Exception as e:
+            item.exception = e
+        return item
+
+    def collect_annotations(self, item):
+        for annotator_key, annotator in item.test.get_annotators().items():
+            annotator_request = annotator.translate_request(item.prompt_with_context(), item.completion())
+            cache_key = annotator_request.model_dump_json(exclude_none=True)
+            self._debug(f"looking for {cache_key} in cache")
+            if cache_key in self.cache:
+                self._debug(f"cache entry found")
+                annotator_response = self.cache[cache_key]
+            else:
+                self._debug(f"cache entry not found; processing and saving")
+                annotator_response = annotator.annotate(annotator_request)
+                self.cache[cache_key] = annotator_response
+
+            annotation = annotator.translate_response(annotator_request, annotator_response)
+            item.annotations[annotator_key] = annotation
+        item.test.measure_quality(item)
+
+
+class TestRunResultsCollector(Sink):
+
+    def __init__(self, test_run: TestRun):
+        super().__init__()
+        self.test_run = test_run
+
+    def handle_item(self, item) -> None:
+        self.test_run.add_finished_item(item)
+
+
+class RunnerBase:
+    def __init__(self, data_dir: pathlib.Path):
+        self.debug = False
+        self.data_dir = data_dir
+        self.secrets = None
+        self.suts = []
+        self.max_items = 10
+        self.thread_count = 1
+
+    def _check_ready_to_run(self):
+        if not self.secrets:
+            raise ValueError("must set secrets")
+
+        if not self.suts:
+            raise ValueError("must call add_sut() at least once")
+
+
+class TestRunner(RunnerBase):
+
+    def __init__(self, data_dir: pathlib.Path):
+        super().__init__(data_dir)
+        self.tests = []
+
+    def add_sut(self, sut: ModelGaugeSut):
+        self.suts.append(sut)
+
+    def add_test(self, test: PromptResponseTest):
+        self.tests.append(test)
+
+    def _check_ready_to_run(self):
+        super()._check_ready_to_run()
+        if not self.tests:
+            raise ValueError("must call add_test() at least once")
+
+    def run(self, run_object=None) -> TestRun:
+        self._check_ready_to_run()
+        run = run_object or TestRun(self)
```

Comment:
This is just a personal preference: usually I try to avoid naming a variable the same as an enclosing function, especially in python or where the symbol is used in different places with different type (e.g. `run()` the bound method on `pipeline`, `run` the method of this object, `run` the run object, etc.

Reply:
Fair enough. I'll rename it to test_run.